DNR: Parsing of translated content blocker rules should be done off the main thread

b-weinstein opened the pull request:
#### cf3fc160de0cc6098114087d485b0f989123108d
<pre>
DNR: Parsing of translated content blocker rules should be done off the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=294679">https://bugs.webkit.org/show_bug.cgi?id=294679</a>
<a href="https://rdar.apple.com/153200218">rdar://153200218</a>

Reviewed by Alex Christensen.

The only part that isn&apos;t thread safe about parsing of the content blocker rules is parsing the CSS
selectors used in display-none rules. However, these will currently never be present in translated rules from
DNR. If we add support for turning DNR rules into content blocker CSS display-none rules, we will need
to revisit this. This PR adds a variable indicating whether or not CSS selectors will be expected in parsing
all the way down the stack. If a CSS selector is seen when it isn&apos;t expected, an assert will be hit and
the rule will be ignored.

A better fix here would be to make the CSS parsing thread safe, and that is tracked by:
<a href="https://rdar.apple.com/153748194">rdar://153748194</a> (CSS selector validation in content blocker rule list parsing should be thread safe)

* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::ContentRuleListStore::compileContentRuleList):
(API::ContentRuleListStore::compileContentRuleListFile):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules):

Canonical link: <a href="https://commits.webkit.org/296408@main">https://commits.webkit.org/296408@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcbc1765a937f14f37497aa899af7c9c1620708a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113613 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110367 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36615 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82317 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111352 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22804 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97643 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62752 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22220 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58340 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92174 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15835 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116734 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35458 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26121 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91344 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35831 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91144 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23227 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/36043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13800 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31201 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35361 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/40898 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35072 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38424 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36755 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -583,21 +583,32 @@ void ContentRuleListStore::getAvailableContentRuleListIdentifiers(CompletionHand
+            return completionHandler(nullptr, parsedRules.error());
+    }
+
+    protectedCompileQueue()->dispatch([protectedThis = Ref { *this }, filePath = WTFMove(filePath).isolatedCopy(), identifier = WTFMove(identifier).isolatedCopy(), json = WTFMove(json).isolatedCopy(), parsedRules = crossThreadCopy(WTFMove(parsedRules).value()), storePath = m_storePath.isolatedCopy(), completionHandler = WTFMove(completionHandler), cssSelectorsAllowed = cssSelectorsAllowed] () mut ...
```
This keeps the parsed runs and the JSON alive at the same time, which is a significant memory regression in a very memory-tight place.

b-weinstein replied:
I think this should keep the same memory behavior as before.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
```diff
@@ -51,6 +51,11 @@ class ContentRuleListStore final : public ObjectImpl<Object::Type::ContentRuleLi
         RemoveFailed
     };
 
+    enum class CSSSelectorsAllowed {
```
nit : bool

achristensen07 reviewed (changes requested):
I think we should just fix the selector parser instead of doing this to fix half the problem.

xeenon commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -583,21 +583,32 @@ void ContentRuleListStore::getAvailableContentRuleListIdentifiers(CompletionHand
+            return completionHandler(nullptr, parsedRules.error());
+    }
+
+    protectedCompileQueue()->dispatch([protectedThis = Ref { *this }, filePath = WTFMove(filePath).isolatedCopy(), identifier = WTFMove(identifier).isolatedCopy(), json = WTFMove(json).isolatedCopy(), parsedRules = crossThreadCopy(WTFMove(parsedRules).value()), storePath = m_storePath.isolatedCopy(), completionHandler = WTFMove(completionHandler), cssSelectorsAllowed = cssSelectorsAllowed] () mut ...
```
```suggestion
    protectedCompileQueue()->dispatch([protectedThis = Ref { *this }, filePath = WTFMove(filePath).isolatedCopy(), identifier = WTFMove(identifier).isolatedCopy(), json = WTFMove(json).isolatedCopy(), parsedRules = crossThreadCopy(WTFMove(parsedRules).value()), storePath = m_storePath.isolatedCopy(), completionHandler = WTFMove(completionHandler), cssSelectorsAllowed] () mutable {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4795,7 +4795,7 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
                 }
 
-                API::ContentRuleListStore::defaultStoreSingleton().compileContentRuleListFile(declarativeNetRequestContentRuleListFilePath(), uniqueIdentifier().isolatedCopy(), String(webKitRules), [this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), hashOfWebKitRules](RefPtr<API::ContentRuleList> ruleList, std::error_code error) mutable {
+                API::ContentRuleListStore::defaultStoreSingleton().compileContentRuleListFile(declarativeNetRequestContentRuleListFilePath(), uniqueIdentifier().isolatedCopy(), String(webKitRules), API::ContentRuleListStore::CSSSelectorsAllowed::NotAllowed, [this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), hashOfWebKitRules](RefPtr<API::ContentRuleList> ruleLis ...
```
This might come back to bite us if / when we allow dNR to contain cosmetic rules.

b-weinstein replied:
I tried to mention that in the commit message:

`The only part that isn't thread safe about parsing of the content blocker rules is parsing the CSS
selectors used in display-none rules. However, these will never be present in translated rules from
DNR.`

But I can be more clear about it.

xeenon replied:
Yeah, _never_ is strong here since it might be _soon_ we have to change this.

b-weinstein replied:
Ah, got it. I meant `currently never`, but I can make this a bit more clear.

b-weinstein committed:
DNR: Parsing of translated content blocker rules should be done off the main thread
https://bugs.webkit.org/show_bug.cgi?id=294679
rdar://153200218

Reviewed by Alex Christensen.

The only part that isn't thread safe about parsing of the content blocker rules is parsing the CSS
selectors used in display-none rules. However, these will currently never be present in translated rules from
DNR. If we add support for turning DNR rules into content blocker CSS display-none rules, we will need
to revisit this. This PR adds a variable indicating whether or not CSS selectors will be expected in parsing
all the way down the stack. If a CSS selector is seen when it isn't expected, an assert will be hit and
the rule will be ignored.

A better fix here would be to make the CSS parsing thread safe, and that is tracked by:
rdar://153748194 (CSS selector validation in content blocker rule list parsing should be thread safe)

* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::ContentRuleListStore::compileContentRuleList):
(API::ContentRuleListStore::compileContentRuleListFile):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules):

Canonical link: https://commits.webkit.org/296408@main
